### PR TITLE
Add Day_Of_Year Function To OpenSearch

### PR DESCRIPTION
**Pull request by GabeFernandez310**
### Description
Adds the `day_of_year` function to the SQL plugin. It behaves like the `dayofyear` function which already exists in the new engine, and both are aligned with [MySQL](https://www.w3schools.com/sql/func_mysql_dayofyear.asp)
 
### Issues Resolved
#722 
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [X] New functionality has user manual doc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 1]
```diff
@@ -479,6 +479,29 @@ public void dayOfYear() {
...
+    assertEquals(INTEGER, expression.type());
+    assertEquals("day_of_year(DATE '2020-08-07')", expression.toString());
+    assertEquals(integerValue(220), eval(expression));
+
+    expression = DSL.day_of_year(DSL.literal("2020-08-07"));
+    assertEquals(INTEGER, expression.type());
+    assertEquals("day_of_year(\"2020-08-07\")", expression.toString());
+    assertEquals(integerValue(220), eval(expression));
+
+    expression = DSL.day_of_year(DSL.literal("2020-08-07 01:02:03"));
```
Comment (MitchellGale-BitQuill) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:499]: What if you give it an invalid string (random characters, just time etc.)

File: docs/user/dql/functions.rst [hunk 2]
```diff
@@ -1458,6 +1458,14 @@ Example::
...
     +---------------------------------+
 
+    os> SELECT DAY_OF_YEAR(DATE('2020-08-26'))
+    fetched rows / total rows = 1/1
+    +-----------------------------------+
+    | DAY_OF_YEAR(DATE('2020-08-26'))   |
+    |-----------------------------------|
+    | 239                               |
+    +-----------------------------------+
+
```
Comment (MitchellGale-BitQuill) [inline docs/user/dql/functions.rst:1468]: Can we add samples for datetime objects and show that it can get dayofyear with those too please.

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java [hunk 3]
```diff
@@ -225,6 +225,17 @@ public void testDayOfYear() throws IOException {
...
+  @Test
+  public void testDay_Of_Year() throws IOException {
+    JSONObject result = executeQuery("select day_of_year(date('2020-09-16'))");
+    verifySchema(result, schema("day_of_year(date('2020-09-16'))", null, "integer"));
+    verifyDataRows(result, rows(260));
+
+    result = executeQuery("select day_of_year('2020-09-16')");
+    verifySchema(result, schema("day_of_year('2020-09-16')", null, "integer"));
+    verifyDataRows(result, rows(260));
+  }
```
Comment (MitchellGale-BitQuill) [inline integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java:237]: Can we please add a datetime object passed to day_of_year.

**Review by MitchellGale-BitQuill — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 1 above)
Comment (acarbonetto) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:499]: Think edge cases: 29th or 30th day of February, the year 1200, the year 4000, 13th month, etc...

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java (hunk 3 above)
Comment (acarbonetto) [reply integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java:237]: Also add a time object passed to day_of_year (does it convert to datetime first?)

**Review by acarbonetto — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 1 above)
Comment (acarbonetto) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:499]: for something fun: get a value of 366 (on a leap year) since that's the maximum value.

File: docs/user/dql/functions.rst [hunk 4]
```diff
@@ -1442,7 +1442,7 @@ DAYOFYEAR
 Description
 >>>>>>>>>>>
 
-Usage:  dayofyear(date) returns the day of the year for date, in the range 1 to 366.
+Usage:  dayofyear(date) returns the day of the year for date, in the range 1 to 366. Also supports day_of_year(date) as alternate syntax.
```
Comment (Yury-Fridlyand) [inline docs/user/dql/functions.rst:1445]: Consider creating a section for `day_of_year` with one line description and reference to `dayofyear`. RST renderer adds an entry into table of content for each section automatically.

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java:237]: Try with all datetime types. `TIME` depends on https://github.com/opensearch-project/sql/pull/1047 though.

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 5]
```diff
@@ -319,8 +320,8 @@ private DefaultFunctionResolver dayOfWeek() {
    * DAYOFYEAR(STRING/DATE/DATETIME/TIMESTAMP).
    * return the day of the year for date (1-366).
    */
-  private DefaultFunctionResolver dayOfYear() {
-    return define(BuiltinFunctionName.DAYOFYEAR.getName(),
+  private DefaultFunctionResolver dayOfYear(BuiltinFunctionName dayOfYear) {
+    return define(dayOfYear.getName(),
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:324]: Same as https://github.com/Bit-Quill/opensearch-project-sql/pull/175#discussion_r1031842704
Function should accept `TIME` argument.

**Review by Yury-Fridlyand — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 5 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:324]:
```
mysql> select dayofyear(curtime());
+----------------------+
| dayofyear(curtime()) |
+----------------------+
|                  329 |
+----------------------+
1 row in set (0.00 sec)
```

Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:324]: Will be accomplished in future work
https://bitquill.atlassian.net/browse/AOS-1266

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 6]
```diff
@@ -479,6 +480,76 @@ public void dayOfYear() {
...
+    assertEquals(INTEGER, expression.type());
+    assertEquals("day_of_year(\"2020-08-07 01:02:03\")", expression.toString());
+    assertEquals(integerValue(220), eval(expression));
+
+    //28th of Feb
+    test_day_of_year("2020-02-28", 59);
+
+    //29th of Feb during leap year
+    test_day_of_year("2020-02-29 23:59:59", 60);
+    test_day_of_year("2020-02-29", 60);
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:514]: Try using `assertAll` or split into different test cases

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 1 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:504]: Fixed in 09e5a3637655c6deb1f65a9282291c8cce0e4a4e and https://github.com/Bit-Quill/opensearch-project-sql/pull/173/commits/6c4d02c85ec5eccbb9eaed11952fc18d32eb29a2

File: docs/user/dql/functions.rst (hunk 2 above)
Comment (GabeFernandez310) [reply docs/user/dql/functions.rst:1500]: Added in https://github.com/Bit-Quill/opensearch-project-sql/pull/173/commits/226c1a58d51cc544925bb59cc65f9bbf38863105 and https://github.com/Bit-Quill/opensearch-project-sql/pull/173/commits/f1ed83261c96515f29fae59c500fee844cf23820

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java (hunk 3 above)
Comment (GabeFernandez310) [reply integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFunctionIT.java:245]: Added in https://github.com/Bit-Quill/opensearch-project-sql/pull/173/commits/4b09fb86c531a619961d6d6691268e97948e67d9

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 6 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java]: Fixed in https://github.com/Bit-Quill/opensearch-project-sql/pull/173/commits/9c3c1f18a471900bea402048557ac187443e4763

**Review by Yury-Fridlyand — APPROVED** [review]

**Review by margarit-h — APPROVED** [review]

**Review by MitchellGale-BitQuill — APPROVED** [review]

**Review by acarbonetto — APPROVED** [review]
LGTM
